File: dsa in c/quick_sort.c
#include <stdio.h>
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}
int partition(int arr[], int low, int high)
{
    int i = low;
    int j = high;
    int pivot = arr[low];
    while (i < j)
    {
        while (arr[i] <= pivot && i < high)
        {
            i++;
        }
        while (arr[j] > pivot && j > low)
        {
            j--;
        }

        if (i < j)
        {
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[low], &arr[j]);
    return j;
}

void quick_sort(int arr[], int low, int high)
{
    if (low < high)
    {

        int p_idx = partition(arr, low, high);
        quick_sort(arr, low, p_idx - 1);
        quick_sort(arr, p_idx + 1, high);
    }
}

int main()
{
    int arr[] = {4, 6, 2, 9, 1};
    int n = sizeof(arr) / sizeof(arr[0]);

    quick_sort(arr, 0, n - 1);

    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);

    return 0;
}

File: dsa in python/quick_sort.py
def partition(nums, low, high):
    pivot = nums[low]
    i = low
    j = high

    while i < j:
        while i <= high - 1 and nums[i] <= pivot:
            i += 1
        while j >= low + 1 and nums[j] >= pivot:
            j -= 1
        if i < j:
            nums[i], nums[j] = nums[j], nums[i]

    nums[low], nums[j] = nums[j], nums[low]
    return j

def quick_sort(nums, low, high):
    if low < high:  # base case
        p_index = partition(nums, low, high)
        quick_sort(nums, low, p_index - 1)
        quick_sort(nums, p_index + 1, high)

nums = [4, 1, 7, 6, 3, 2, 8]
n = len(nums)
quick_sort(nums, 0, n - 1)
print(nums)

# in best and averge case tc=o(nlogn)
# in worst case tc=o(n^2)
# sc=o(1)